resources/type-system/inference.md

```diff
@@ -1117,6 +1134,26 @@ succintly, the syntax of Dart is extended to allow the following forms:
   any loss of functionality, provided they are not trying to construct a proof
   of soundness._
 
+In addition, the following forms are added to allow constructor invocations,
+dynamic method invocations, function object invocations, instance method
+invocations, and static/toplevel method invocations to be distinguished:
+
+- `@DYNAMIC_INVOKE(m_0.id<T_1, T_2, ...>(n_1: m_1, n_2: m_2, ...))`
+
+- `@FUNCTION_INVOKE(m_0.call<T_1, T_2, ...>(n_1: m_1, n_2: m_2, ...))`
+
+- `@INSTANCE_INVOKE(m_0.id<T_1, T_2, ...>(n_1: m_1, n_2: m_2, ...))`
+
+- `@STATIC_INVOKE(f<T_1, T_2, ...>(n_1: m_1, n_2: m_2, ...))`
```

Comment:
What is the syntactic category of `f`?  Is it only an identifier?  Or prefixed identifier perhaps?  Probably not an expression?

Reply:
Hmm, this is a situation where my choice of notation really fails to capture what’s in my head.

My intention is for `@STATIC_INVOKE` to be the spec counterpart of the kernel’s `StaticInvocation` class (https://github.com/dart-lang/sdk/blob/f1febd2cf3bf754ae641eb3fda8ee885b8bda5e5/pkg/kernel/lib/ast.dart#L6531), with `f` corresponding to `StaticInvocation.targetReference`, which has type `Reference`. A kernel `Reference` is an indirection mechanism allowing different parts of the kernel representation of a program to refer to one another unambiguously. In practice it’s serialized as a sequence of strings separated by `::`, the first of which is a URI (e.g. `dart:core::int::tryParse`), but that’s an unimportant implementation detail. What’s important is that `StaticInvocation.targetReference` is some way of referring to unambiguously to a top level function or static method defined in a library somewhere in the user’s program.

Putting on my spec writer’s hat, I guess if we want to be really formal about it, we could say that `f` is either a pair `(URI, name)` referring to a top level function named `name` in the library located at `URI`, or a triple `(URI, name1, name2)`, where `name1` is the name of a class, mixin, enum, or extension, and `name2` is the name of a static method inside the thing named by `name1`. But I don’t really want to be this formal because it will make my examples really unwieldy (e.g. I want to say `@STATIC_INVOKE(print(s))` rather than `@STATIC_INVOKE(('dart:core', 'print')(s))`).

Also, even `(URI, name1, name2)` isn’t good enough to uniquely identify a static method inside an unnamed extension, since a single library could have multiple unnamed extensions containing static methods with the same name. The kernel representation deals with this by giving every unnamed extension a synthetic name, but that feels like an implementation detail of the kernel that doesn’t belong in the spec.

What I *really* want to say is: “gentle reader, please think of `f` using whatever representation suits you for unambiguously referring to a static method or top level function that exists somewhere in the program being type inferred. Maybe that’s a combination of a URI and some strings, with some synthetic naming magic to disambiguate methods in unnamed extensions. Maybe it’s a unique integer identifier you assign to each static method or top level function in the user’s program. For my purposes in this document, I’m going to try to stick to examples where it’s clear from context what I’m referring to, like `@STATIC_INVOKE(print(s))` (which clearly refers to the top level function `print` in `dart:core` or `@STATIC_INVOKE(int.tryParse(s))` (which clearly refers to the static method `tryParse` in the class `int` in `dart:core`).”

I’m not really sure how to express that idea in the sort of formal style that we typically use for specifications. Any suggestions?

Reply:
Why do you want to move to the semantic layer here?  In the examples above, you just use the elaborated expression as the target.  Why not something similar here?  That is, you are starting with an expression (or maybe prefixed identifier?) `f`, why not just continue to use that in the meta-syntax?